### PR TITLE
feat(ingestion): durable SCHEDULED dispatch with lease fencing and recovery

### PR DESCRIPTION
**Pull request by jay77721**
### Summary / Motivation

Legacy ingestion scheduling relied on optimistic `CREATED → RUNNING` CAS transitions, unconstrained broker redeliveries (`NATS MaxDeliver16` / `AckWait60s`), and a startup-only `CREATED` reconciler. This design suffered from three critical reliability flaws:
1. **Permanently Stranded Tasks**: Crashed or OOM-killed workers left tasks in `RUNNING` status indefinitely with no automatic recovery.
2. **Ambiguous Dispatch Intent & Data Loss**: Broker publishing errors or stream restarts dropped `CREATED` tasks or triggered destructive cleanups.
3. **Unbounded Poison Redeliveries**: Faulty/panicking tasks repeatedly failed and looped through redelivery queues without a circuit breaker.

This PR establishes an **Intent-First, Durable DB Lease & Lifecycle-Owned Recovery** model. Scheduling intent is persisted as `SCHEDULED` prior to MQ publishing, workers claim tasks via atomic database leases (`claim_token` + `claim_expires_at`), dual heartbeats renew active leases, terminal transitions are fenced by tokens to prevent split-brain writes, and a rate-limited periodic reconciler recovers expired leases with bounded poison-task capping. Fixes #19042.

### Key Changes

- **Schema & Status (`constants`, `entity`, `migration`)**:
  - Added `SCHEDULED` status (`common.SCHEDULED` and `document.run = "5"`).
  - Added scheduling & lease columns (`scheduled_at`, `last_dispatched_at`, `claim_token`, `claim_expires_at`, `lease_recovery_attempt`) backed by composite indexes `(status, last_dispatched_at, id)` and `(status, claim_expires_at, id)` with online-DDL migration guides and `TABLE_SCHEMA = DATABASE()` scoping.
- **DAO Lifecycle (`dao/ingestion`)**:
  - Implemented atomic primitives: `Schedule`, `TryClaim`, `TouchClaim`, `FinalizeClaim`, `ReleaseClaim`, `RecoverExpiredClaim`, `ReleaseExpiredClaim`, `StopExpiredClaim`, `TryReserveDispatch`, `ListScheduledForDispatch` (keyset pagination), and `ListExpiredClaims`.
- **Task Service (`service/ingestion_task_service`)**:
  - Transitioned to intent-first `CreateAndEnqueue` / `Schedule` (`SCHEDULED`).
  - Managed atomic worker claims (`SCHEDULED → RUNNING`) with synchronized `document.run = RUNNING(1)` mapping.
  - Deduped concurrent dispatches via `TryReserveDispatch` CAS on `last_dispatched_at` using shared `IngestionDispatchGracePeriod`.
- **Worker & Reconciler (`ingestion/service`)**:
  - **Dual Heartbeat**: Background goroutine calls `InProgress()` (NATS) and `TouchClaim` (DB) every heartbeat interval; auto-cancels context upon lease loss or DB unavailability.
  - **Fenced Terminal Writes**: `completeTask`, `markFailed`, and `markStopped` verify `claim_token` through `FinalizeClaim` to reject stale worker writes.
  - **Lifecycle-Owned Reconciler**: Background reconciler runs alongside worker lifecycle with a `100/s` pace ticker; recovers expired claims, marks poison tasks as `FAILED` after reaching `maxLeaseRecoveryAttempts`, and redrives un-dispatched `SCHEDULED` tasks.
- **Configuration & Server (`ingestor_config`, `service_conf.yaml`, `ragflow_server`)**:
  - Added `claim_ttl` (default `15s`) and `max_lease_recovery_attempts` (default `3`) with strict startup validation (`claim_ttl > 3 * heartbeat`, `heartbeat < 5s`).
- **Web & API (`handler/document`, `web/*`)**:
  - Mapped `document.run = "5"` to `SCHEDULED` badge; updated polling (`isParserRunning`) and filter normalization across Go and React frontends.

### Implementation Details

- **Intent-First Ordering**: Tasks are persisted as `SCHEDULED` before publishing. If NATS is temporarily unavailable, the task remains safely recorded in MySQL and is automatically dispatched on the next reconciler sweep after `IngestionDispatchGracePeriod`.
- **Token CAS Fencing**: `FinalizeClaim` requires matching `claim_token` and `claim_expires_at > now`. If a worker's lease expires due to a long GC pause, its subsequent terminal write affects 0 rows, preventing corrupted state writes over a new owner.
- **Poison Capping**: `RecoverExpiredClaim` increments `lease_recovery_attempt`. Once attempts exceed the configured threshold, the task transitions to `FAILED` and `markPoisonProgress` appends failure details to the document.
- **Keyset Pagination & Concurrent Dedup**: Reconciler scans use keyset cursors `(last_dispatched_at, id)` and advance past transient publish failures; `TryReserveDispatch WHERE last_dispatched_at = ?` CAS guarantees only one ingestor publishes a given task per cycle.

### How to Test

```bash
# Run targeted unit and integration test suites
bash build.sh --test ./internal/dao
bash build.sh --test ./internal/service
bash build.sh --test ./internal/service/document
bash build.sh --test ./internal/ingestion/service
bash build.sh --test ./internal/server/config
bash build.sh --test ./internal/handler -run "TestMapRunStatus|TestNormalizeRunStatus"
```

### CODE REVIEW

**Comment by jay77721** [comment]
### Review Verdict: Request Changes (minor, non-blocking) — otherwise LGTM

This PR correctly closes the P1 reliability gap: `SCHEDULED` intent + `15s` DB lease (`claim_token/claim_expires_at`) + `2s` `TouchClaim WHERE token && expiry>now` + `FinalizeClaim WHERE token` fencing + periodic `reconciler` (converge `CREATED→SCHEDULED` / recover expired `RUNNING/STOPPING` / dispatch `SCHEDULED` with `keyset` + `100/s jitter`) + `lease_recovery_attempt=3` poison cap.

**What is solved thoroughly:** `MaxDeliver16` without DLQ / permanent `RUNNING` stranded / `CREATED` ambiguity / cross-process double-claim and stale owner overwriting terminal state. The design reuses the proven `knowledge_compile/scheduler.go:569/594` lease pattern and is well tested (8 new DAO cases, heartbeat loss-cancel, poison-FAILED with document `-1`, lost-claim fencing).

**What is intentionally not solved in P1:** Cross-store fencing. P1 only fences `ingestion_task` terminal writes; `ES BulkInsert` / `MinIO` / `nav` writes can still complete after `Cancel()` — honestly documented in the plan as P2 `lease_epoch` work.

**Required fixes before merge (4 small items):**
1. **Hot DDL:** `ALTER ingestion_task ADD 5 cols + 2 composite indexes` needs `gh-ost`/`ALGORITHM=INPLACE` window and rollback `DROP INDEX` in deployment notes.
2. **Clock:** `TouchClaim`/`TryClaim` use `time.Now()` for `claim_expires_at`. On multi-node NTP skew this can misjudge expiry. Use `DB NOW(3)` or widen `TTL 15s→20s`.
3. **Hard-coded cap:** `maxLeaseRecoveryAttempts=3` is hard-coded in `ingestion_service.go:970`. Expose via `ingestor_config.go` and truncate `markPoisonProgress` message to `3000` (`progress_sink.go:38`).
4. **Migration gap:** `validateTransition` removes `FAILED/STOPPED→CREATED` and `MarkCompleted/MarkFailed`. Rolling upgrade must converge remaining `CREATED` rows to `SCHEDULED` before switching workers — add example `UPDATE ... SET status='SCHEDULED' WHERE status='CREATED'` to notes.

With those addressed, **LGTM**. P2 `monotonic epoch + outbox` can follow separately.

*Reviewed offline from `/tmp/pr19042.patch` (read-only, no worktree pollution).*

**Comment by jay77721** [comment]
**Follow-up: make `claimTTL` / `lease_recovery_max` configurable**

Current:
- `defaultClaimTTL = 15s` and `maxLeaseRecoveryAttempts = 3` are hard-coded in `internal/ingestion/service/ingestion_service.go:908/970`
- `heartbeatInterval = 2s` is already configurable via `ingestor_config.go` (`MaxConcurrentWorkers` path) but `claimTTL` is not

Suggestion:
```go
// internal/server/config/ingestor_config.go
type IngestorConfig struct {
    MaxConcurrentWorkers int `mapstructure:"max_concurrent_workers"`
    CompilerPoolSize     int `mapstructure:"compiler_pool_size"`
    ClaimTTL             time.Duration `mapstructure:"claim_ttl"` // default 15s, must be > heartbeatInterval * 3
    LeaseRecoveryMax     int           `mapstructure:"lease_recovery_max"` // default 3
}
```
- Wire via `cmd/ragflow_server.go:566 SetClaimTTL/SetRecoveryMax` (or `ingestion.NewIngestor(..., claimTTL, recoveryMax)`)
- Validate `claimTTL > heartbeatInterval * 3` on startup, otherwise fail fast
- This allows tuning for long-tail `30m` pipelines (e.g., `20s` for NTP skew) vs. fast poison detection (`2` for dev)

*Keeps the current 15s/3 defaults, just makes them ops-tunable without code change.*

**Comment by jay77721** [comment]
## Code Review Summary

**Verdict: LGTM with minor changes**

This PR correctly implements the P1 plan from RafactorPlan.md — durable SCHEDULED + owner lease + token CAS + periodic reconciler. The "Schedule before Publish" order is safer than the alternative (publish-then-mark), and it eliminates the old code's dangerous behavior of deleting new tasks / rolling back terminal state on publish failure.

### What works well
- Clean 5 DAO atomic operations (Schedule/TryClaim/TouchClaim/FinalizeClaim/RecoverExpiredClaim) following the proven knowledge_compile lease pattern
- Token CAS fencing prevents stale workers from writing terminal state
- Periodic reconciler (not just startup) with keyset pagination and rate limiting
- Bounded lease recovery (`lease_recovery_attempt` + `maxLeaseRecoveryAttempts=3`) prevents infinite redelivery
- P1/P2 boundary is explicit — no over-engineering
- Test coverage is thorough: concurrent claim, lease expiry, poison doc, keyset paging, claim-loss rejection, heartbeat renewal

### Issues to address before/after merge

**Minor (non-blocking):**
1. `validateTransition` has dead code — `SCHEDULED → RUNNING` is never called (TryClaim goes through DAO directly), and `STOPPING → STOPPED` is also unused. Clean up or add a comment.
2. `TouchClaim` has a redundant `claim_token <> ''` condition (already implied by `claim_token = ?`).
3. `convergeCreatedTasks` lacks rate limiting, unlike `dispatchScheduledTask` (which has `reconcileRate=100/s`). A large backlog of legacy CREATED rows could flood publishes on first reconcile.

**Confirm before merge:**
1. Does the frontend handle `document.run = "5"` (TaskStatusSchedule) introduced by `markDocumentScheduled`?
2. Is the product semantics acceptable where a task enters SCHEDULED but may wait up to `reconcileInterval` (15s) before being dispatched?
3. Is `leaseTTL=15s` acceptable under DB jitter? A brief 2-3s DB stall could expire the lease and kill a 30-min parse. The heartbeat distinguishes DB error from lease loss via `ClaimExpiresAt <= now`, but please verify this edge case.

### Test gaps to fill
- The old `TestStartRunningLeavesTerminalDocumentUntouched` (terminal doc counters not clobbered) was removed without an equivalent replacement.
- Add a rolling-upgrade compatibility test: old code writes CREATED, new code only accepts SCHEDULED.
- Add bulk `convergeCreatedTasks` rate-limit verification.

### Bottom line
Significantly improves system reliability. Core mechanism is sound. Fix the minor points above and this is ready to merge.

**Comment by jay77721** [comment]
## Code Review Summary: PR #19042 (`feat(ingestion): add durable task scheduling leases`)

### Overall Assessment
This is a high-quality, well-architected PR that thoroughly addresses the historical ingestion scheduling reliability issues in RAGFlow. The core design (P1 scope) is robust, cohesive, and adheres closely to our single-path convergence principles:
1. **Durable Intent-First Dispatch**: Persisting `SCHEDULED` state in DB before publishing to NATS avoids task loss or rollback on broker publish errors.
2. **Distributed DB Leases & Single-Owner Fencing**: Using atomic CAS `TryClaim` with UUID tokens (`claim_token`, 15s TTL) and requiring token match on `FinalizeClaim` effectively prevents multi-worker conflicts and stale zombie overwrites.
3. **Dual Heartbeats & Cooperative Cancellation**: Combining broker `InProgress` renewals with DB `TouchClaim` and immediate context cancellation on lease loss prevents wasted compute.
4. **Periodic Reconciler & Poison Pill Bounding**: Replacing startup-only reconciliation with a lifecycle-owned periodic scan, while bounding lease recovery via `lease_recovery_attempt` (failing poison documents after threshold), prevents crash loops.

---

### Findings & Actionable Fixes

#### 1. [Crash / Bug] Guard against `dao.DB == nil` in `startHeartbeat`
- **Location**: `internal/ingestion/service/ingestion_service.go:1182`
- **Issue**: In lightweight unit test harnesses where `dao.DB` is nil or uninitialized, the heartbeat goroutine calling `e.ingestionTaskSvc.TouchClaim(...)` invokes `dao.DB.WithContext(...)`, which panics with a `nil pointer dereference` (`SIGSEGV`).
- **Suggested Fix**:
  ```diff
  --- a/internal/ingestion/service/ingestion_service.go
  +++ b/internal/ingestion/service/ingestion_service.go
  @@ -1175,6 +1175,9 @@ func (e *Ingestor) startHeartbeat(taskCtx *taskpkg.TaskContext) func() {
   				if taskCtx.IngestionTask.ClaimToken == "" {
   					continue
   				}
  +				if dao.DB == nil {
  +					continue
  +				}
   				leaseCtx, cancel := context.WithTimeout(context.WithoutCancel(taskCtx.Ctx), 5*time.Second)
   				alive, err := e.ingestionTaskSvc.TouchClaim(leaseCtx, taskCtx.IngestionTask.ID, taskCtx.IngestionTask.ClaimToken, e.claimTTL)
   				cancel()
  ```

#### 2. [Test Regression] Update legacy `CREATED` status in `document_test.go` test helper
- **Location**: `internal/service/document/document_test.go:568`
- **Issue**: `insertTestIngestionTask` still defaults to `common.CREATED`. Since `RequestStop` and `CancelDocParse` have been updated to expect `SCHEDULED`, `TestStopParseDocuments_UnfinishedTask` and `TestIngest_CancelUnstartedWithInFlightTask` fail assertion checks.
- **Suggested Fix**:
  ```diff
  --- a/internal/service/document/document_test.go
  +++ b/internal/service/document/document_test.go
  @@ -566,3 +566,3 @@ func insertTestIngestionTask(t *testing.T, id, userID, docID, datasetID string) {
   	t.Helper()
  -	insertTestIngestionTaskWithStatus(t, id, userID, docID, datasetID, common.CREATED)
  +	insertTestIngestionTaskWithStatus(t, id, userID, docID, datasetID, common.SCHEDULED)
   }
  ```

---

### Conclusion
**LGTM with the two minor fixes above.** Outstanding work on solidifying the ingestion scheduling layer!

**Comment by jay77721** [comment]
## Follow-up Review: Latest changes (fix commit)

The new commit addresses several important points. Here's my review of what changed:

### Strengths of the new changes

1. **Config-driven lease settings** (`conf/service_conf.yaml`, `ingestor_config.go`): `claim_ttl: 15s` and `max_lease_recovery_attempts: 3` with validation that `claim_ttl > 3 * heartbeat_interval`. Clean and safe.

2. **Frontend SCHEDULE status** (`handler/document.go`): Added `"5" → SCHEDULE` mapping in `mapRunStatus`, `normalizeRunStatusFilter`, and `invalidRunStatuses`. This correctly handles the new `document.run = "5"` state that `markDocumentScheduled` writes.

3. **DAO safety improvements**: `ReleaseExpiredClaim`, `RecoverExpiredClaim`, `StopExpiredClaim`, `FinalizeClaim`, `ListExpiredClaims` now all require `claim_token <> '' AND claim_expires_at > 0`. This prevents accidental manipulation of unleased rows — important for the legacy RUNNING rows that have empty lease fields.

4. **Legacy RUNNING recovery** (`ListLegacyRunningTasks` + `RecoverLegacyRunningTask`): Handles old RUNNING rows left by pre-lease writers. The staleness predicate (`claim_token = '' AND claim_expires_at <= 0 AND update_time < olderThan`) correctly identifies rows that cannot be recovered by the active-lease scan.

5. **Concurrent dispatch reservation** (`TryReserveDispatch`): Uses `WHERE last_dispatched_at = ?` CAS so concurrent ingestor reconcilers publish a SCHEDULED task only once. Verified by `TestConcurrentIngestorsReserveScheduledDispatchOnce`.

6. **Rate limiting on all reconcilers**: `convergeCreatedTasks`, `recoverLegacyRunningTasks`, `recoverExpiredClaims`, `dispatchScheduledTasks` all now use `pace.Ticker(time.Second / reconcileRate)`. This addresses my earlier concern about bulk CREATED convergence flooding publishes.

7. **DB-unavailable heartbeat handling**: `startHeartbeat` now cancels the context when `dao.DB == nil`, with a test `TestStartHeartbeatStopsWhenDatabaseIsUnavailable`.

8. **Progress log Unicode safety**: `limitProgressLog` uses rune slicing to avoid splitting multi-byte characters.

9. **settleClaimLoss recursion guard**: The comment explaining why `markStopped` is called directly (not via `settleClaimLoss`) for STOPPING prevents infinite recursion. Good defensive design.

### Remaining observations (non-blocking)

1. **`validateTransition` still has unused branches**: `SCHEDULED → RUNNING` is never called (TryClaim uses DAO directly), and `STOPPING → STOPPED` is also unused (markStopped uses FinalizeClaim). These are now effectively dead code. Consider removing or annotating.

2. **leaseTTL=15s with 2s heartbeat**: The heartbeat distinguishes DB error from lease loss via `ClaimExpiresAt <= now`, which is correct. But under sustained DB latency (e.g., 3-4s), a healthy worker's lease can still expire. The `claim_ttl > 3 * heartbeat` validation helps, but please monitor lease expiry metrics in production.

3. **`dispatchGracePeriod = 2 minutes`**: This means a SCHEDULED task won't be re-dispatched until 2 minutes after its last dispatch. Combined with `reconcileInterval = 15s`, this is reasonable, but worth noting: if a SCHEDULED task's message is lost (e.g., stream purge), it takes up to 2 minutes to recover.

### Verdict

The fix commit significantly strengthens the PR. The config layer, frontend status mapping, DAO safety predicates, legacy row handling, and concurrent dispatch reservation are all well-designed. Ready to merge after addressing the minor `validateTransition` cleanup.

**Comment by jay77721** [comment]
## Re-review of latest HEAD (4 commits, 4081 lines)

**Verdict: LGTM — Ready to merge ✅**

The two new commits (`test(document): update test helpers to use SCHEDULED status` and `fix(review): address coderabbit feedback`) resolve nearly all previously identified issues and significantly harden the implementation.

### What's new and notable

| Improvement | Detail |
|---|---|
| Config validation | `SetLeaseConfig` enforces `claimTTL > 3×heartbeat` |
| Migration robustness | `migrateIngestionTaskSchedulingIndexes` adds columns first, uses `TABLE_SCHEMA = DATABASE()` to avoid cross-db false matches |
| Legacy RUNNING recovery | `ListLegacyRunningTasks` + `RecoverLegacyRunningTask` handles old unleased RUNNING rows from pre-lease writers |
| Concurrent dispatch dedup | `TryReserveDispatch` (`WHERE last_dispatched_at = ?` CAS) ensures N ingestors publish a SCHEDULED task only once |
| Rate limiting on all reconcilers | `convergeCreatedTasks`, `recoverLegacyRunningTasks`, `recoverExpiredClaims`, `dispatchScheduledTasks` all pace at 100/s |
| DB-unavailable heartbeat | Cancels context when `dao.DB == nil`, with test coverage |
| Frontend SCHEDULE status | `document.run = "5"` → `SCHEDULE` in `mapRunStatus`, `normalizeRunStatusFilter`, `invalidRunStatuses` |
| Progress log safety | `limitProgressLog` uses rune slicing to avoid splitting multi-byte chars |
| Poison document marking | `markPoisonProgress` writes FAILED + progress message after recovery cap exceeded |
| Recursion guard | `settleClaimLoss` calls `finalizeClaim` directly for STOPPING (not `markStopped`), with comment explaining infinite-recursion prevention |

### Test coverage: excellent

12+ new test cases covering: concurrent claim, lease expiry, bounded recovery, poison doc, keyset paging, dispatch dedup, configured TTL, DB unavailable, lost-claim rejection, progress message length, frontend status mapping.

### Only remaining nit (non-blocking)

`validateTransition` retains `SCHEDULED → RUNNING` and `STOPPING → STOPPED` branches that are no longer called (TryClaim/FinalizeClaim use DAO CAS directly). Dead code — can be cleaned up in a follow-up.

### Bottom line

This is a well-designed, thoroughly tested reliability improvement. The Schedule-before-Publish ordering eliminates the old code's task-loss-on-publish-failure defect. Token CAS fencing prevents stale-worker writes. The periodic reconciler with rate limiting and bounded recovery handles the full failure spectrum. Merge-ready.

**Review by Haruko386 — APPROVED** [review]
LGTM
